tsfc/coffee_mode.py

```diff
@@ -109,6 +108,49 @@ def monomial_sum_to_expression(monomial_sum):
     return make_sum(indexsums)
 
 
+def solve_ip(size, is_feasible, compare):
+    """Solve a 0-1 integer programming problem. The algorithm tries to set each
+    variable to 1 recursively, and stop the recursion early by comparing with
+    the optimal solution at entry. At worst case this is 2^N (as this is a
+    NP-hard problem), but recursions can be trimmed as soon as a reasonable
+    solution have been found. One potential improvement is to keep track of
+    violated constraints at each step, and only try to set the variables which
+    could help these constraints in the next step. This will help find decent
+    solutions faster with the additional overhead of maintaining a list of
+    violated constraints and finding helpful variables.
```

Comment:
The purpose of documentation is to describe what one needs to know to figure out whether some function solves his particular problem and how to use it. As with everything else, it is supposed to be minimal but sufficient. Since the purpose of documentation is not to duplicate the code in English prose, thorough description of implementation details should be avoided in the documentation. If a brief description of the implementation is necessary to help the understanding of the code, it should happen in comments rather than in the documentation.

From "One potential improvement...": Ideation about potential improvements definitely does not belong to the documentation. Make it a comment if you find it necessary to be in the source code.